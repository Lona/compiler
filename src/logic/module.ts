import fs from 'fs'
import path from 'path'
import {
  decodeDocument,
  decodeLogic,
  extractProgramFromAST,
  LogicAST as AST,
  MDXAST,
} from '@lona/serialization'
import { IFS, match } from 'buffs'
import { evaluate, EvaluationContext } from './evaluation'
import { createNamespace, mergeNamespaces, Namespace } from './namespace'
import { createScopeContext, mergeScopes, Scope } from './scope'
import { createUnificationContext, TypeChecker } from './typeChecker'
import { Substitution, unify } from './typeUnifier'
import { joinPrograms, makeProgram } from './ast'

export type LogicFile = {
  isLibrary: boolean
  sourcePath: string
  rootNode: AST.TopLevelDeclarations
  mdxContent: MDXAST.Content[]
}

export type ModuleContext = {
  componentFiles: LogicFile[]
  libraryFiles: LogicFile[]
  documentFiles: LogicFile[]
  logicFiles: LogicFile[]
  sourceFiles: LogicFile[]
  namespace: Namespace
  scope: Scope
  typeChecker: TypeChecker
  substitution: Substitution
  evaluationContext: EvaluationContext
  isValidProgram: boolean
}

export function createModule(
  workspaceFs: IFS,
  workspacePath: string
): ModuleContext {
  const libraryFiles: LogicFile[] = libraryFilePaths().map(sourcePath => ({
    isLibrary: true,
    sourcePath,
    // Always read library files from the real FS
    rootNode: decodeLogic(
      fs.readFileSync(sourcePath, 'utf8')
    ) as AST.TopLevelDeclarations,
    mdxContent: [],
  }))

  const componentFiles: LogicFile[] = componentFilePaths(
    workspaceFs,
    workspacePath
  ).map(sourcePath => ({
    isLibrary: false,
    sourcePath,
    rootNode: decodeLogic(
      workspaceFs.readFileSync(sourcePath, 'utf8')
    ) as AST.TopLevelDeclarations,
    mdxContent: [],
  }))

  const logicFiles: LogicFile[] = logicFilePaths(
    workspaceFs,
    workspacePath
  ).map(sourcePath => ({
    isLibrary: false,
    sourcePath,
    rootNode: decodeLogic(
      workspaceFs.readFileSync(sourcePath, 'utf8')
    ) as AST.TopLevelDeclarations,
    mdxContent: [],
  }))

  const documentFiles: LogicFile[] = documentFilePaths(
    workspaceFs,
    workspacePath
  ).map(sourcePath => {
    const decoded = decodeDocument(workspaceFs.readFileSync(sourcePath, 'utf8'))

    return {
      isLibrary: false,
      sourcePath,
      rootNode: extractProgramFromAST(decoded),
      mdxContent: decoded.children,
    }
  })

  const files: LogicFile[] = [
    ...libraryFiles,
    ...componentFiles,
    ...documentFiles,
    ...logicFiles,
  ]

  const namespace: Namespace = mergeNamespaces(
    files.map(logicFile => createNamespace(logicFile.rootNode))
  )

  const scope: Scope = mergeScopes(
    files.map(logicFile =>
      createScopeContext(logicFile.rootNode, namespace, undefined, console)
    )
  )

  const programNode = joinPrograms(
    files.map(logicFile => makeProgram(logicFile.rootNode))
  )

  const typeChecker = createUnificationContext(programNode, scope, console)

  const substitution = unify(typeChecker.constraints, console)

  const evaluationContext = evaluate(
    programNode,
    namespace,
    scope,
    typeChecker,
    substitution,
    console
  )

  return {
    componentFiles,
    libraryFiles,
    documentFiles,
    logicFiles,
    get sourceFiles() {
      return [...documentFiles, ...logicFiles]
    },
    namespace,
    scope,
    typeChecker,
    substitution,
    evaluationContext,
    isValidProgram:
      scope.undefinedIdentifierExpressions.size === 0 &&
      scope.undefinedMemberExpressions.size === 0 &&
      scope.undefinedTypeIdentifiers.size === 0,
  }
}

function componentFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, { includePatterns: ['**/*.cmp'] }).map(file =>
    path.join(workspacePath, file)
  )
}

function logicFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, {
    includePatterns: ['**/*.logic'],
  }).map(file => path.join(workspacePath, file))
}

function documentFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, { includePatterns: ['**/*.md'] }).map(file =>
    path.join(workspacePath, file)
  )
}

export function libraryFilePaths(): string[] {
  return match(fs, STANDARD_LIBRARY_PATH, {
    includePatterns: ['**/*.logic'],
  }).map(file => path.join(STANDARD_LIBRARY_PATH, file))
}

export const STANDARD_LIBRARY_PATH = path.join(__dirname, '../../static/logic')
