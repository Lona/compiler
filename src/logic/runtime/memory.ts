import { StaticType } from '../staticType'
import { Value } from './value'

export type RecordMemory = { [key: string]: Value }

export type DefaultArguments = {
  [key: string]: [StaticType, Value | void]
}

export type FuncImplementation = (args: RecordMemory) => Memory

export type FuncMemory = {
  f: FuncImplementation
  defaultArguments: DefaultArguments
}

export type Memory =
  | { type: 'unit' }
  | { type: 'bool'; value: boolean }
  | { type: 'number'; value: number }
  | { type: 'string'; value: string }
  | { type: 'array'; value: Value[] }
  | { type: 'enum'; value: string; data: Value[] }
  | { type: 'record'; value: RecordMemory }
  | {
      type: 'function'
      value: FuncMemory
    }

export const unit = (): Memory => ({ type: 'unit' })

export const bool = (value: boolean): Memory => ({ type: 'bool', value })

export const number = (value: number): Memory => ({ type: 'number', value })

export const string = (value: string): Memory => ({ type: 'string', value })

export const array = (value: Value[]): Memory => ({ type: 'array', value })
