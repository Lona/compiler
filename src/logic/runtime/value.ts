import * as StaticType from '../staticType'
import * as Memory from './memory'

export type Value = {
  type: StaticType.StaticType
  memory: Memory.Memory
}

export namespace Encode {
  export const unit = (): Value => ({
    type: StaticType.unit,
    memory: Memory.unit(),
  })

  export const bool = (value: boolean): Value => ({
    type: StaticType.bool,
    memory: Memory.bool(value),
  })

  export const number = (value: number): Value => ({
    type: StaticType.number,
    memory: Memory.number(value),
  })

  export const string = (value: string): Value => ({
    type: StaticType.string,
    memory: Memory.string(value),
  })

  export const color = (value: string): Value => ({
    type: StaticType.color,
    memory: {
      type: 'record',
      value: {
        value: {
          type: StaticType.string,
          memory: Memory.string(value),
        },
      },
    },
  })

  export const array = (
    elementType: StaticType.StaticType,
    values: Value[]
  ) => ({
    type: elementType,
    memory: Memory.array(values),
  })
}

export namespace Decode {
  export const string = ({ type, memory }: Value): string | undefined => {
    if (
      type.type === 'constructor' &&
      type.name === 'String' &&
      memory.type === 'string'
    ) {
      return memory.value
    }
  }
  export const number = ({ type, memory }: Value): number | undefined => {
    if (
      type.type === 'constructor' &&
      type.name === 'Number' &&
      memory.type === 'number'
    ) {
      return memory.value
    }
  }

  export const color = ({ type, memory }: Value): string | undefined => {
    if (
      type.type === 'constructor' &&
      type.name === 'Color' &&
      memory.type === 'record'
    ) {
      const colorValue = memory.value['value']
      return string(colorValue)
    }
  }

  export const optional = ({ type, memory }: Value): Value | undefined => {
    if (
      type.type === 'constructor' &&
      type.name === 'Optional' &&
      memory.type === 'enum' &&
      memory.value === 'value'
    ) {
      return memory.data[0]
    }
  }

  export const fontWeightToNumberMapping: { [key: string]: string } = {
    ultraLight: '100',
    thin: '200',
    light: '300',
    regular: '400',
    medium: '500',
    semibold: '600',
    bold: '700',
    heavy: '800',
    black: '900',
  }

  export const fontNumberToWeightMapping: {
    [key: string]: string
  } = Object.fromEntries(
    Object.entries(fontWeightToNumberMapping).map(([key, value]) => [
      value,
      key,
    ])
  )

  export const fontWeight = ({ type, memory }: Value): string | undefined => {
    if (
      type.type === 'constructor' &&
      type.name === 'FontWeight' &&
      memory.type === 'enum'
    ) {
      return fontWeightToNumberMapping[memory.value]
    }
  }

  export type EvaluatedShadow = {
    x: number
    y: number
    blur: number
    radius: number
    color: string
  }

  export const shadow = ({
    type,
    memory,
  }: Value): EvaluatedShadow | undefined => {
    if (
      type.type === 'constructor' &&
      type.name === 'Shadow' &&
      memory.type === 'record'
    ) {
      return {
        x: Decode.number(memory.value['x']) ?? 0,
        y: Decode.number(memory.value['y']) ?? 0,
        blur: Decode.number(memory.value['blur']) ?? 0,
        radius: Decode.number(memory.value['radius']) ?? 0,
        color: Decode.color(memory.value['color']) ?? 'black',
      }
    }
  }

  export type EvaluatedTextStyle = {
    fontName?: string
    fontFamily?: string
    fontWeight?: number
    fontSize?: number
    lineHeight?: number
    letterSpacing?: number
    color?: string
  }

  export const textStyle = ({
    type,
    memory,
  }: Value): EvaluatedTextStyle | undefined => {
    if (
      type.type === 'constructor' &&
      type.name === 'TextStyle' &&
      memory.type === 'record'
    ) {
      const fontName = Decode.optional(memory.value['fontName'])
      const fontFamily = Decode.optional(memory.value['fontFamily'])
      const fontWeight = memory.value['fontWeight']
      const fontSize = Decode.optional(memory.value['fontSize'])
      const lineHeight = Decode.optional(memory.value['lineHeight'])
      const letterSpacing = Decode.optional(memory.value['letterSpacing'])
      const color = Decode.optional(memory.value['color'])

      return {
        ...(fontName && { fontName: Decode.string(fontName) }),
        ...(fontFamily && { fontFamily: Decode.string(fontFamily) }),
        ...(fontWeight && {
          fontWeight: Number(Decode.fontWeight(fontWeight)),
        }),
        ...(fontSize && { fontSize: Decode.number(fontSize) }),
        ...(lineHeight && { lineHeight: Decode.number(lineHeight) }),
        ...(letterSpacing && { letterSpacing: Decode.number(letterSpacing) }),
        ...(color && { color: Decode.color(color) }),
      }
    }
  }
}
