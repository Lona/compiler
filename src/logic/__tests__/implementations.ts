import { createFs } from 'buffs'
import { createModule } from '../module'
import { RecordMemory } from '../runtime/memory'
import { Encode } from '../runtime/value'

describe('Logic / Implementations', () => {
  it('implements the standard library', () => {
    const source = createFs({ 'lona.json': JSON.stringify({}) })
    const module = createModule(source, '/')

    // Detect missing functions in the global namespace
    const missingImplementations = Object.entries(
      module.namespace.values
    ).flatMap(([qualifiedNameString, id]) => {
      const value = module.evaluationContext.evaluate(id)

      if (!value) {
        throw new Error(`Problem evaluating ${qualifiedNameString}`)
      }

      // Find all functions
      if (value.memory.type === 'function') {
        const { defaultArguments, f } = value.memory.value

        const args: RecordMemory = Object.fromEntries(
          Object.entries(defaultArguments).map(([name, [_, val]]) => {
            return [name, typeof val === 'undefined' ? Encode.unit() : val]
          })
        )

        // Call each function to check if it exists
        const result = f(args)

        // A function that returns unit most likely doesn't exist
        if (result.type === 'unit') {
          return [qualifiedNameString]
        }
      }

      return []
    })

    // Assert missing implementations. This should be an empty array before committing.
    expect(missingImplementations).toMatchSnapshot()
  })
})
