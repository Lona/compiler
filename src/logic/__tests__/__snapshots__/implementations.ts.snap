// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logic / Implementations implements the standard library 1`] = `
Array [
  "Color.setHue",
  "Color.setSaturation",
  "Color.setLightness",
  "Color.fromHSL",
  "Boolean.or",
  "Boolean.and",
  "Number.range",
  "String.concat",
  "Array.at",
]
`;
